Add auto refresh function

tokuchan-pg opened the pull request:
# WHAT
設定したインターバル（１０秒）中に他のユーザーが投稿したメッセージが、メッセージ一覧ページに非同期で表示されるように実装する。（ページの自動更新）
SQLの発行回数を減らすため、投稿済みのメッセージの全てを自動更新するのではなく、インターバル中に投稿されたメッセージ（前回のインターバル終了時点からの差分）のみを非同期で取得し、表示する。

# WHY
アプリケーションのリアルタイム性を高め、よりよいUXを提供するため

## インターバル中に投稿されたメッセージをAPI（messages#index）で取得するロジック
**STEP１**
各メッセージを表示しているdivタグ（.chat-main__body--message）に、メッセージのidをカスタムデータ属性として持たせる
![2017-05-05 13 15 53](https://cloud.githubusercontent.com/assets/25572309/25733531/01079b48-3195-11e7-9676-0b5b0f4aa4a9.png)

↓
**STEP２**
javascriptにて、最後に投稿されたメッセージのdivタグからSTEP１で設定したidを取得し、API（messages#index）に送る。
まだメッセージが投稿されていない場合は、0をidとして代入する。（undefinedが代入されるのを防ぎ、APIでインターバル中に投稿された全てのメッセージを取得するため）
![2017-05-05 13 17 50](https://cloud.githubusercontent.com/assets/25572309/25733571/4cb9bda0-3195-11e7-99a7-37e3264e776d.png)

↓
**STEP３**
API（messages#index）にて、idの値がajaxで送られてきたidよりも大きいメッセージを取得し、javascriptに返す
![2017-05-05 13 22 29](https://cloud.githubusercontent.com/assets/25572309/25733610/f617d90e-3195-11e7-896f-a8aefe47b1a5.png)

↓
**STEP４**
APIから返ってきたメッセージ（配列）をforEachを使ってHTMLを組成する
（配列が空の場合（インターバル中に何も投稿されなかった場合）はforEachが起動せず、HTMLも組成されない）


## 画面の挙動（上の画面にメッセージを入力）
[![https://gyazo.com/d0c3c2437b35598ce4f6940737009ed8](https://i.gyazo.com/d0c3c2437b35598ce4f6940737009ed8.gif)](https://gyazo.com/d0c3c2437b35598ce4f6940737009ed8)

tokuchan-pg committed:
add message id as a data property of each messages when HTML is requested

Message id is needed as a data property to get messages posted in
interval. This message id will be sent to  API(messages#index) by Ajax.
In the API, it assigns messages into @messages_add of which the ids are
bigger than the id(lastMessageID) sent from message.js.

tokuchan-pg committed:
add auto refresh function

Using messages#index as a API to return messages which are posted in
the interval(10 sec).

ninooo29 commented on app/controllers/messages_controller.rb:
```diff
@@ -4,6 +4,14 @@ class MessagesController < ApplicationController
+      format.json do
+        # @messages_addはHTMLリクエストに対しては生成不要なので、format.json内に入れ子にして定義
+        # idが、ajaxで送られてきた最後のメッセージのidよりも大きいメッセージを取得
+        @messages_add = Message.where(group_id: params[:group_id]).where('id > ?', params[:lastMessageID])
```
Message.where(group_id: params[:group_id])のところは、現在選択しているgroupのmessageのような形でとってきましょう。

tokuchan-pg replied:
下記のように修正いたしました！
`@messages_add = @group.messages.where('id > ?', params[:lastMessageID])`

ninooo29 commented on app/assets/javascripts/message.js:
```diff
@@ -68,4 +71,38 @@ $(document).on("turbolinks:load", function() {
+    })
+
+    .done(function(data) {
+      data.forEach(function(message_add){
```
doneのときの処理の前にdataの数が0でないときという条件を追加しましょう。

ninooo29 commented:
メッセージが送信されたとき自動でスクロールされる記述はされていますでしょうか？

tokuchan-pg committed:
used association to get messages from current group

tokuchan-pg committed:
build html only if data(return value from API) has more than one element

tokuchan-pg committed:
enabled chat-main page to scroll to see overflowed messages and groups

tokuchan-pg committed:
added auto scroll function to see the latest message when posting a new message or when the page is refreshed automatically

tokuchan-pg committed:
activate auto reload function only while the user is staying at message#index view

if the URL of current page matches the path of group#index, activate
ajax. if not, clear the interval.

tokuchan-pg commented:
## 以下の修正をいたしました
- 自動更新用のAPIからの返り値である配列の要素数が１以上のときにのみHTMLを組成するように設定
![2017-05-05 21 47 57](https://cloud.githubusercontent.com/assets/25572309/25746034/88010cc8-31dc-11e7-9f22-076efd3b0576.png)

- cssファイルにてoverflow-y: scroll;を指定し、画面に収まらないメッセージとグループをスクロールで表示できるように設定

- 最新のメッセージが画面内に表示されるように、自動スクロールを設定（メッセージ送信時 ＆ 自動更新時の２箇所にて設定）
![2017-05-05 21 58 49](https://cloud.githubusercontent.com/assets/25572309/25746393/2df87db8-31de-11e7-8092-d1e5400f2c56.png)

- messages#indexのページでのみsetInterval（自動更新のためのajaxのコードを内包している）が有効になり、それ以外のページではclearIntervalでsetIntervalを停止できるように設定。条件判定は、閲覧中のページのURLを取得し、messages#indexのパスと正規表現でマッチングさせることで実現。（※コードの全体像は↓の画像をご参照いただけたらと思います）


## setInterval内のコード
![2017-05-05 21 38 42](https://cloud.githubusercontent.com/assets/25572309/25745820/b9eec816-31db-11e7-8f21-9de8ad776525.png)

ninooo29 commented on app/assets/javascripts/message.js:
```diff
@@ -68,4 +78,51 @@ $(document).on("turbolinks:load", function() {
+      // Turbolinksを止めないためにfalseを返しておく
+      return false;
+
+    } else {
```
(location.pathname.match(/\/groups\/\d+\/messages/))からelseまでの処理が長いのでメソッド化して呼び出しましょう。

tokuchan-pg replied:
自動更新のためのAjax部分をメソッド化（getLatestMessages();）いたしました！

ninooo29 commented on app/assets/javascripts/message.js:
```diff
@@ -68,4 +78,51 @@ $(document).on("turbolinks:load", function() {
+    clearInterval(autoReload);
+    }
+
+  }, 10000);
```
setInterval(function()がここまで続いているのだとおもいますが、もう少しメソッドで分けるなどして簡潔にしましょう。

tokuchan-pg replied:
自動更新のためのAjax部分をメソッド化（getLatestMessages();）したことで、setIntervalの中のコードも簡潔できました。

tokuchan-pg committed:
made the whole Ajax codes for auto refresh function as a getLatestMessages()

tokuchan-pg commented:
#### ↓の要領で自動更新のための一連のAjaxコードをメソッド化し（getLatestMessages();）、コード全体を整理いたしました

![evernote snapshot 20170506 123931](https://cloud.githubusercontent.com/assets/25572309/25769678/0934f33a-325b-11e7-9e33-1583f3d357f1.png)
